models/classes/class.QtiTestService.php

```diff
@@ -1454,6 +1483,16 @@ private function getQtiPackageImportPreprocessing()
         return $this->getServiceLocator()->get(QtiPackageImportPreprocessing::SERVICE_ID);
     }
 
+    private function getQtiItemService(): QtiItemService
```

Comment:
in theory with above comments this method shouldnt be necessary, neither  getElementReferencesExtractor()
As a nitpick, consider dependency injection rather than using a container if possible

Reply:
I'd too prefer to use DI here, but this being a legacy service makes it harder (and I'm trying to prevent adding to many changes to this file). Some places are already instantiating this as `QtiTestService::singleton()`.

Moreover, having these in separate methods makes it easier to use DI from tests in the future (but I'm also refraining from doing that here to prevent adding too many changes): The code calls the service getters, and in the future the getters may be rewritten to pick an injected service if available, or call the singleton otherwise .

```php
private function getQtiItemService(): QtiItemService
{
    if (!($this->qtiItemService instanceof QtiItemService)) {
        $this->qtiItemService = QtiItemService::singleton();
    }

    return $this->qtiItemService;
}
```

Reply:
@hectoras not needed, but if you are trying to improve that (I agree with it is not necessary), we can easily do that on DI container using `factory` method. Please, do not create a private property with cache :)

Reply:
Changed in 87436f8d5 :ok_hand:

Reply:
I could not find any container reference for this `Service`. Are you sure this is working as expected? (calling the `singleton` inside the container cache file).

Reply:
It is working fine for me

![image](https://github.com/oat-sa/extension-tao-testqti/assets/339792/c8356135-fd79-4bd4-a70e-a110af4e946f)

I was assuming the LegacyServiceGateway was doing the right thing and calling the `singleton()` method. I'll recheck and (most probably) rollback to use `QtiItemService::singleton()` again

Reply:
Thanks! It is all good!